src/lib.rs

```diff
@@ -35,10 +37,23 @@ impl KVMi {
                           unsafe extern fn(*const kvmi_qemu2introspector,
                                            *mut kvmi_introspector2qemu,
                                            *mut c_void) -> c_int);
-        let ctx = null_mut();
-        unsafe {
-            kvmi_sys::kvmi_init_unix_socket(socket_path.as_ptr(), accept_db, hsk_cb, ctx);
+        let mut kvmi = KVMi {
+            ctx: null_mut(),
+            connect_condvar: Condvar::new(),
+        };
+        let cb_ctx: *mut c_void = &kvmi;
+        // kvmi_dom = NULL
+        let guard = Mutex::new(false);
+        let lock = guard.lock().unwrap();
+        let res: *mut c_void = unsafe {
+            kvmi_sys::kvmi_init_unix_socket(socket_path.as_ptr(), accept_db, hsk_cb, cb_ctx)
```

Comment:
Hey, we should be careful at `kvmi_init_unix_socket`. Is its first param a null-terminated (i.e. `C` string)? If yes, then passing `socket_path.as_ptr()` should be very dangerous because a `Rust`'s string needs not to be null-terminated.

Reply:
How do you suggest transforming the Rust String into a C null terminated one ?

Reply:
Something likes:

```rust
let socket_path = CString::new(socket_path.clone()).unwrap();
kvmi_sys::kvmi_init_unix_socket(socket_path.as_ptr(), ....
```

I think we need to either clone the original or consume it, because appending a zero into an existing `String` is hard (note that `String` is utf8-encoded) and not always possible.

Reply:
Fixed.